src/storage/garbage_collector.cpp

```diff
@@ -129,54 +243,279 @@ void GarbageCollector::ProcessDeferredActions() {
-  // is newest-to-oldest sorted.
+  }
+
+  if (interval_length > 1) {
+    // Create the undo record by a second traversal through the records to be compacted.
+    UndoRecord *compacted_undo_record = CreateUndoRecord(start_record, next);
+    // Copy the varlen attributes of the compacted record
+    CopyVarlen(compacted_undo_record);
+    // Link the compacted record to the version chain.
+    LinkCompactedUndoRecord(start_record, &curr, next, compacted_undo_record);
+  } else if (interval_length == 1) {
+    if (transaction::TransactionUtil::NewerThan(active_txns->back(), curr->Timestamp().load())) {
+      // Exited out of previous loop such that the last record is not visible to any transaction
+      UnlinkUndoRecordVersion(start_record->Next());
+      start_record->Next().store(nullptr);
+    }
+  }
+  // Set interval length to 0
+  EndCompaction(&interval_length);
+
+  // active_trans_iter ends but there are still elements in the version chain. Can GC everything below
   curr->Next().store(nullptr);
+  while (next != nullptr) {
+    // Unlink next
+    UnlinkUndoRecordVersion(next);
+    // Move next pointer ahead
+    next = next->Next();
+  }
+}
 
-  // If the head of the version chain was not committed, it could have been aborted and requires a retry.
-  if (curr == version_ptr && !transaction::TransactionUtil::Committed(version_ptr->Timestamp().load()) &&
-      table->AtomicallyReadVersionPtr(slot, accessor) != version_ptr)
-    TruncateVersionChain(table, slot, oldest);
+void GarbageCollector::UnlinkUndoRecordVersion(UndoRecord *const undo_record) {
+  TERRIER_ASSERT(!undo_record->txnptr_.IsNull(), "Table should not be NULL here");
+  ReclaimSlotIfDeleted(undo_record);
+  ReclaimVarlen(undo_record);
+  // Mark the record as fully processed
+  undo_record->txnptr_.Put(nullptr);
 }
 
 void GarbageCollector::ReclaimSlotIfDeleted(UndoRecord *const undo_record) const {
   if (undo_record->Type() == DeltaRecordType::DELETE) undo_record->Table()->accessor_.Deallocate(undo_record->Slot());
 }
 
-void GarbageCollector::ReclaimBufferIfVarlen(transaction::TransactionContext *const txn,
-                                             UndoRecord *const undo_record) const {
+void GarbageCollector::BeginCompaction(UndoRecord **start_record_ptr, UndoRecord *curr, UndoRecord *next,
+                                       uint32_t *interval_length_ptr) {
+  col_set_.clear();
+  // Case 1: next is of type UPDATE
+  //         Standard undo record compaction begins
+  // Case 2: next is of type INSERT
+  //         This can only happen when a transaction is sandwiched between INSERT and another Delta record
+  //         INSERT is the last record in the version chain
+  // Case 3: next is of type DELETE
+  //         DELETE can only be the newest record in the version chain
+  TERRIER_ASSERT(next->Type() != DeltaRecordType::DELETE, "Delete cannot be compacted");
```

Comment:
It is great that all callers of this method guarantee that `next` is not `nullptr`. Maybe document this as an assumption in the header or check again here explicitly? This will certainly help future students.